### PR TITLE
New Docker Setup

### PR DESCRIPTION
**Pull request by cody-vibe**
Let's discuss about it before merging.

Let me know if you have any problems with the instructions from the README.md file.
The main ideas behind the change were:
- Vagrant is a quick access to a generic Linux VM with necessary tools installed but running on a linux host or a VM not created by vagrant should be possible and documented.
- The /work and /vagrant directories can be used to share the code between host and guest although the code can be moved anywhere inside the VM.
- The production docker image contains the code it needs inside.
- The development docker image is built on top of the production one and code is shared inside it.
- Some other small improvements I discovered.

### CODE REVIEW

**Comment by marius-balteanu** [comment]
@sdwolf, what do you think if we drop this repo and we use the official repo? https://github.com/docker-library/redmine

**Comment by cody-vibe** [comment]
They are not equivalent, the official repo just gives a way to create the redmine docker image while this one handles building, development / production setup, and orchestration.

File: docs/linux.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    source ~/.bash_profile
+    ```
+
+4. Clone this repository:
+
+    ``` bash
+    git clone https://github.com/marius-balteanu/redmine_docker.git
+    ```
+
+5. Go inside the `redmine_docker/project` directory and clone the redmine repository:
```
Comment (marius-balteanu) [inline docs/linux.md]: It's not clear where is the `redmine_docker/project` directory. It should be created or already exists?

Comment (marius-balteanu) [inline docs/linux.md]: my bad.

**Comment by marius-balteanu** [comment]
1. We should install vim by default.
2. When I run the command `docker-compose -f docker/docker-compose.yml up -d` I receive the following error: 
```
Pulling repository docker.io/library/my_redmine_development
ERROR: Error: image library/my_redmine_development:latest not found
```
2. Why we don't run steps 2, 3, 4 and 5 automatically on vagrant provision?

**Comment by cody-vibe** [comment]
1) Yes I will add vim
2) You need to run the development build command first (step 8 second command)
3) Steps 2 and 3 are generic enough to be added to the Vagrantfile as well, 4 needs to say you can also use the directory in the /vagrant share and for 5+ I'm thinking of using Ansible to run them.

**Comment by marius-balteanu** [comment]
2. Error output:
```
Building web_development
Step 1 : FROM my_redmine
Pulling repository docker.io/library/my_redmine
ERROR: Service 'web_development' failed to build: Error: image library/my_redmine:latest not found
```

**Comment by marius-balteanu** [comment]
I tried again from zero and I receive the same error message (from my previous comment).

**Comment by cody-vibe** [comment]
@marius-balteanu I'l take a look this evening.

**Comment by cody-vibe** [comment]
For production:

1. Have this in your `secrets/env.list`:
    ```shell
    RAILS_DATABASE_NAME=redmine
    RAILS_DATABASE_HOST=mysql
    RAILS_DATABASE_USER=root
    RAILS_DATABASE_PASSWORD=root
    RAILS_ENV=production
    SECRET_TOKEN=thisisarandomstringthatmustbereplacedwithaproppersecrettoken
    ```
    Most important line is: `RAILS_ENV=production`

2. Copy the custom files:
    ```bash
    \cp -f ./config/Gemfile.local ./redmine/
    \cp -f ./config/additional_environment.rb ./redmine/config/
    \cp -f ./config/database.yml ./redmine/config/
    \cp -f ./config/secret_token.rb ./redmine/config/initializers
    ```

3. Build the production image:
    ```bash
    docker-compose -f docker/docker-compose.build.yml build
    ```

4. Bring the production services up:
    ```bash
    docker-compose -f docker/docker-compose.production.yml up -d
    ```

5. Inspect the services:
    ```bash
    docker-compose -f docker/docker-compose.production.yml ps
    ```
    You should see this:
    ```
         Name                   Command               State          Ports
    ------------------------------------------------------------------------------
    docker_mysql_1   docker-entrypoint.sh mysqld      Up      3306/tcp
    docker_web_1     /entry bundle exec puma -w ...   Up      0.0.0.0:80->9292/tcp
    ```

6. Create and migrate the database:
    ```
    docker-exec docker_web_1 /entry bundle exec rake db:create db:migrate
    ```

7. Visit `redmine.local/` or `0.0.0.0` and there you go!

**Comment by cody-vibe** [comment]
For development, in addition to the production setup above you do:

1. Stop the production environment:
   ```bash
    docker-compose -f docker/docker-compose.production.yml stop
   ```

2. Change the `RAILS_ENV` variable to `development` inside `secrets/env.list`:
    ```shell
    RAILS_ENV=development
    ```

2. Build the development image:
    ```bash
    docker-compose -f docker/docker-compose.build-dev.yml build
    ```

3. Bring the development services up:
    ```bash
    docker-compose -f docker/docker-compose.yml up -d
    ```

4. Inspect the services:
    ```bash
    docker-compose -f docker/docker-compose.yml ps
    ```
    You should see this:
    ```bash
           Name                      Command               State              Ports
    ------------------------------------------------------------------------------------------
    redmine_mysql         docker-entrypoint.sh mysqld      Up      3306/tcp
    redmine_nginx_proxy   /app/docker-entrypoint.sh  ...   Up      443/tcp, 0.0.0.0:80->80/tcp
    redmine_web           /entry bundle exec puma -w ...   Up      9292/tcp
    ```

5. Since they use the same database you don't need to create and migrate it again. Just visit the same url.

**Comment by marius-balteanu** [comment]
It works with the new instructions.  Thanks 👍

**Comment by cody-vibe** [comment]
Glad to hear, I will update the documentation with the above steps for clarity.

**Comment by marius-balteanu** [comment]
@sdwolf, can I merge this?

**Comment by cody-vibe** [comment]
@marius-balteanu let me take another look at it this evening and I will tell you If there is anything else I need done here.

**Comment by cody-vibe** [comment]
There are some breaking changes that I would like to add to this repo that would make cleanup harder later on, I already implemented them somewhere else and I want to add them here as well. 

I will most likely do them over the weekend. In the meantime, tell me if the `Vagrantfile` is OK with the `nfs` plugin or should I restore the samba configuration (you were saying something about performance problems on the `nfs` part for Windows).

**Comment by marius-balteanu** [comment]
The nfs plugin works ok for me.

**Comment by cody-vibe** [comment]
@marius-balteanu can you please check out the instructions in the `host_setup.md` file while I do some more changes in the `Vagrantfile`?

**Comment by cody-vibe** [comment]
OK, this is no longer a WIP. Awaiting your review. Let me know if you encounter any difficulties, especially with the documentation.

**Pull request merged**